src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs

```diff
@@ -28,65 +27,42 @@ public CompareAttributeAdapter(CompareAttribute attribute, IStringLocalizer stri
+            var errorMessage = GetErrorMessage(context);
+            var clientRule = new ModelClientValidationEqualToRule(errorMessage, "*." + Attribute.OtherProperty);
             return new[] { clientRule };
         }
 
-        private static string FormatPropertyForClientValidation(string property)
+        /// <inheritdoc />
+        public override string GetErrorMessage(ModelValidationContextBase validationContext)
         {
-            return "*." + property;
-        }
-
-        private sealed class CompareAttributeWrapper : CompareAttribute
-        {
-            public CompareAttributeWrapper(CompareAttribute attribute)
-                : base(attribute.OtherProperty)
+            if (validationContext == null)
             {
-                // Copy settable properties from wrapped attribute. Don't reset default message accessor (set as
-                // CompareAttribute constructor calls ValidationAttribute constructor) when all properties are null to
-                // preserve default error message. Reset the message accessor when just ErrorMessageResourceType is
-                // non-null to ensure correct InvalidOperationException.
-                if (!string.IsNullOrEmpty(attribute.ErrorMessage) ||
-                    !string.IsNullOrEmpty(attribute.ErrorMessageResourceName) ||
-                    attribute.ErrorMessageResourceType != null)
-                {
-                    ErrorMessage = attribute.ErrorMessage;
-                    ErrorMessageResourceName = attribute.ErrorMessageResourceName;
-                    ErrorMessageResourceType = attribute.ErrorMessageResourceType;
-                }
+                throw new ArgumentNullException(nameof(validationContext));
             }
 
-            public string FormatErrorMessage(ClientModelValidationContext context)
-            {
-                var displayName = context.ModelMetadata.GetDisplayName();
-                return string.Format(CultureInfo.CurrentCulture,
-                                     ErrorMessageString,
-                                     displayName,
-                                     GetOtherPropertyDisplayName(context));
-            }
+            var displayName = validationContext.ModelMetadata.GetDisplayName();
+            var otherPropertyDisplayName = GetOtherPropertyDisplayName(validationContext);
+            return GetErrorMessage(validationContext.ModelMetadata, displayName, otherPropertyDisplayName);
+        }
 
-            private string GetOtherPropertyDisplayName(ClientModelValidationContext context)
+        private string GetOtherPropertyDisplayName(ModelValidationContextBase validationContext)
+        {
+            // The System.ComponentModel.DataAnnotations.CompareAttribute doesn't populate the
+            // OtherPropertyDisplayName until after IsValid() is called. Therefore, by the time we get
```

Comment:
Confusion in old comment: "by the time" is incorrect since client-side `GetErrorMessage()` calls are early, not late. Suggest "at the time".